### ISSUE TITLE
Map and Array Sizes

### ISSUE THREAD

**Issue by drbartling** [cbor/spec-with-errata-fixed#1]
It seems to me that the length of arrays and maps would be better given as a byte size.  This would allow someone reading the CBOR object to jump to the next item after the array or map without having to traverse the entire array.

That is, having the number of elements rather than total size of the array comes at a large performance penalty, and just adds complexity.

```   
| 0xa0..0xb7      | map (0x00..0x17 pairs of data items follow)     |
   |                 |                                                 |
   | 0xb8            | map (one-byte uint8_t for n, and then n pairs   |
   |                 | of data items follow)                           |
```

**Comment by adelyte-chris** [comment]
I have wished for this myself at times, but the debate is ultimately between rapid, efficient scanning and rapid, efficient parsing.

I have come round to CBOR as specced because:
  1. CBOR is intended to be JSON-like, and JSON supports mixed contents for arrays and maps. 
      - Calculating the byte size of an array or map with mixed contents is nontrivial and, for applications that do not require scanning, unhelpful.
      - Calculating the byte size of a typed array with fixed-width contents, like a `Uint8Array`, is trivial, but even fixed-width datatypes like `byte` and `short` and `uint` have valid variable-width CBOR encodings, and therefore appear "mixed".
      - Arrays and maps can be of indefinite length.
  3. The array or map size given as the count of its contents fits with JavaScript `Array.prototype.length`, `Map.prototype.size`, etc.
  4. For particular use cases with optimized scanning needs, a custom tagging scheme with an embedded length can be used. This requires a modification to the CBOR parser, of course, but all custom tags do.

**Comment by drbartling** [comment]
Thanks for the reply,

I'd rather not have the element count at all.

In most cases it will consume the same number of bytes as the open map char + break.  And for long byte stings (as an example) it will consume more memory.  It just adds to the complexity (and program size) when trying to read back the data, since you have to be able to handle both cases (specified size (and all the ways it can be specified) and unspecified (open-break)).

I don't see the benefit of having the specified count option for a number of reasons.  For one, before you request the size of a map or array, you'll need to parse the whole thing anyway and make sure it's valid for the sake of security alone.  Therefore, you will always need to pay the price to traverse the CBOR buffer at least (and hopefully at most) once.

If we must have something, then at least make it something that gives a perf win for scanning or random access actions or something.

**Comment by jtippet** [comment]
Coincidentally, I was just needing the same thing, and I stumbled across this discussion.

I tossed together [a proposal for a new tag](https://github.com/jtippet/cbor-tags/blob/master/homogenous-numeric-array.md) that gives you random access of numeric arrays.  (Basically it's the same idea as ECMAScript's `Uint32Array`).

I'd really appreciate any comments (even ones of the sort "do not do this terrible thing to CBOR").
